dali/pipeline/operator/arg_helper.h

```diff
@@ -15,94 +15,170 @@
-  inline bool IsSet() const noexcept { return is_set_; }
 
-  template <int ndim = DynamicDimensions>
-  inline const TensorView<StorageCPU, T, ndim> Tensor(int sample_index) const {
-    ASSERT(IsInput());
-    return view<T, ndim>(arg_input_[sample_index]);
+template <typename T, int ndim = 0>
+class ArgValue {
+ public:
+  using TLV = TensorListView<StorageCPU, const T, ndim>;
+  using TV = TensorView<StorageCPU, const T, ndim>;
+
+  ArgValue(std::string arg_name, const OpSpec &spec)
+      : arg_name_(std::move(arg_name)) {
+    has_arg_const_ = spec.HasArgument(arg_name_);
+    has_arg_input_ = spec.HasTensorArgument(arg_name_);
+    assert(!(has_arg_const_ && has_arg_input_));
   }
 
-  inline const T &Value() const {
-    return value_;
+  bool IsDefined() const {
+    return has_arg_const_ || has_arg_input_;
   }
 
-  inline const string &Name() const noexcept {
-    return name_;
+  bool IsConstant() const {
+    return has_arg_const_;
   }
 
-  inline const T &operator[](int sample_index) const {
-    if (IsInput()) {
-#if DALI_DEBUG
-      DALI_ENFORCE(sample_index < static_cast<int>(arg_input_->ntensor()));
-#endif
-      return (*arg_input_)[sample_index].data<T>()[0];
+  bool IsArgInput() const {
+    return has_arg_input_;
+  }
+
+  void Acquire(const OpSpec &spec, const ArgumentWorkspace &ws, int nsamples,
+               const TensorShape<ndim> &expected_shape) {
+    if (has_arg_input_) {
+      view_ = view<const T, ndim>(ws.ArgumentInput(arg_name_));
+      DALI_ENFORCE(is_uniform(view_.shape) && expected_shape == view_.shape[0],
+        make_string("Expected uniform shape for argument \"", arg_name_,
+                    "\" but got shape ", view_.shape));
     } else {
-      return  value_;
+      if (ndim == 0) {
+        data_ = {spec.GetArgument<T>(arg_name_)};
+      } else {
+        data_ = spec.GetRepeatedArgument<T>(arg_name_);
+        int64_t len = data_.size();
+        int64_t expected_len = volume(expected_shape);
+        if (len == 1 && expected_len > 1) {
+          data_.resize(expected_len, data_[0]);
+        } else {
+          DALI_ENFORCE(len == volume(expected_shape),
+                       make_string("Argument \"", arg_name_, "\" expected shape ", expected_shape,
+                                   " but got ", len,
+                                   " values, which can't be interpreted as the expected shape."));
+        }
+      }
+      view_ = constant_view(nsamples, data_.data(), expected_shape);
     }
   }
 
-  inline const TensorList<GPUBackend> &AsGPU(cudaStream_t stream) {
-    DALI_ENFORCE(IsInput());
-    if (!gpu_)
-      gpu_.reset(new TensorList<GPUBackend>());
-    if (gpu_dirty_) {
-      gpu_->Copy(*arg_input_, stream);
-      gpu_dirty_ = false;
+  template <typename ShapeFromSizeFn = ArgShapeFromSize<ndim>>
+  void Acquire(const OpSpec &spec, const ArgumentWorkspace &ws, int nsamples,
+               bool enforce_uniform = false, ShapeFromSizeFn &&shape_from_size = {}) {
```

Comment:
Looking at the usage, I would say the `enforce_uniform` should be true by default.